File: demos/matlab/demo_toy_images.m
%% demo TOY IMAGES

clear
addpath(genpath('../../src/GLFMmatlab/'));
addpath(genpath('../../src/Ccode/'));

randn('seed',round(sum(1e5*clock)));
rand('seed',round(sum(1e5*clock)));

%% GENERATIVE MODEL
N = 1000;
s2x = 1;
[data,gT] = generate_toy_images(N,s2x);

%% INITIALIZATION + PARAMETER SETTINGS
Zini = double(rand(N,1)>0.8);
hidden.Z = Zini;

% define params
params.missing = -1;
params.s2Y = 1;     % Variance of the Gaussian prior on the auxiliary variables (pseudoo-observations) Y
params.s2u = .001;
params.s2B = 1;      % Variance of the Gaussian prior of the weigting matrices B
params.alpha = 1;    % Concentration parameter of the IBP
params.Niter = 100; % Number of iterations for the gibbs sampler
params.maxK = 10;
params.bias = 0;

%% INFERENCE
hidden = IBPsampler_run(data, hidden, params);

%% PLOT REAL Vs INFERRED LATENT FEATURE
Kest = size(hidden.B,2);
Zp = eye(Kest);
% compute observations resulting when each of the latent features is active
X_map = IBPsampler_MAP(data.C, Zp, hidden);

figure(1); title('Ground truth');
for k=1:size(gT.B,1)
    subplot(3,3,k);
    imagesc(reshape(gT.B(k,:), [6 6]));
end
hold off;

figure(2); title('Inferred Latent Features');
for k=1:Kest
    subplot(3,3,k);
    imagesc(reshape(X_map(k,:), [6 6]));
end
hold off;

File: src/GLFMR/demo_toyImages.R
# Demo toyImages
# image(m) es el analogo a imagesc en matlab

rm(list=ls())
setwd("~/Documents/Working_papers/FAP_Rpackage/GLFM/src/GLFMR")
source("GLFM_infer.R")
source("GLFM_computeMAP.R")
source("generate_toy_images.R")

# Generative model
N<-200
s2x<-0.5

#Initialisation
data_gen<- generate_toy_images(N,s2x)
data<-list("X"=data_gen$X,"C"=data_gen$C)
#gT<-list(""=)
m0<-matrix(0,nrow=N,ncol=1)
Z <- apply(m0, c(1,2), function(x) sample(c(0,1),1,prob=c(0.8,0.2)))
# define params
alpha <- 2   # Concentration parameter of the IBP
Niter <- 100  # Number of iterations for the Gibbs sampler
maxK <- 10
params<-list("alpha"=alpha,"Niter"=Niter,"maxK"=maxK)
# Inference
output <- GLFM_infer(data, list(Z,params))
Kest<-dim(output$hidden$B)[1]
Zp <-diag(Kest)
X_map <- GLFM_computeMAP(data$C, Zp, output$hidden, output$params,c())
# Plot, ground truths
for(k in 1:dim(data_gen$gTB)[1]){
image(matrix(data_gen$gTB[1,],nrow=6,ncol=6))
}
# Plot inferred 
for(k in 1:Kest){
  image(matrix(X_map[k,],nrow=6,ncol=6))
  readline("Press return to continue")
}

